mac-capture: Improve window capture support on macOS 12.3

Developer-Ecosystem-Engineering opened the pull request:
#### Test Builds
* [Apple Silicon (ARM64)](https://github.com/obsproject/obs-studio/suites/6855299307/artifacts/264738652)
* [Intel (x86_64)](https://github.com/obsproject/obs-studio/suites/6855299307/artifacts/264738653)

*Builds from 09.06.2022 (DD.MM.YYYY)*

### Description
Add a new capture plugin called General Capture that allows for capture of an entire desktop, a single window, or all windows of an application. Performs similarly to the existing macOS Display Capture and includes the ability to hide the cursor. Outperforms Window Capture with significantly higher framerates, with lower CPU/GPU utilization in both OBS and WindowServer. New application capture capabilities allow users to show an application and all of its windows without displaying the content of their desktop.
### Motivation and Context
The current implementation of Window Capture uses `CGWindowListCreateImage` to generate individual frames for window capture. This process requries a significant amount of CPU and GPU utilization from both OBS and WindowServer. ScreenCaptureKit allows General Capture to use the same video plugin style as Display Capture with better performance and lower resource utilization. ScreenCaptureKit provides `IOSurfaceRef`s instead of having to render `CGImage`s on a timer.
Display Capture uses `CGDisplayStream` which provides efficient access to the screen buffer. ScreenCaptureKit allows for the cursor, applications, or windows to be excluded from a display capture. Therefore, General Capture can provide a desktop capture similar to Display Capture with additional features to the existing Display Capture.
Application captures will allow use cases like application tutorials to include only the application in question and the menu bar. This can prevent accidental display of the desktop, notifications, or windows from other applications that may appear on screen. Window Capture allows users to hide other applications and only show the main window of an application, but any additional windows would have to be manually added. Additionally, the menu bar would not be visible in a Window Capture.
<details>
	<summary>Baldur’s Gate 3 Demo</summary>

![Baldur’s Gate 3- Window Capture vs General Capture Comparison](https://user-images.githubusercontent.com/65677710/151424590-4a4ee0fa-3daa-4389-b92b-50beae2e70d8.mp4)
</details>
<details>
	<summary>Shadow of the Tomb Raider</summary>

![Shadow of the Tomb Raider - Window Capture vs General Capture Comparison](https://user-images.githubusercontent.com/65677710/151421432-ca5c615b-4fa4-420c-a298-53255fc63d67.mp4)
</details>
<details>
	<summary>Shadow of the Tomb Raider (25% Speed Playback)</summary>

![Shadow of the Tomb Raider - Window Capture vs General Capture Comparison - Running at 25% speed](https://user-images.githubusercontent.com/65677710/151421310-9907c476-e989-46b3-8438-821036815d04.mp4)
</details>


### How Has This Been Tested
Tested on a 2019 16" Macbook Pro with 2.3 GHz 8-Core i9 and AMD Radeon Pro 5500M graphics and an 8GB M1 Mac Mini on macOS Monetery 12.3.


Games tested include: Dota 2, Shadow of the Tomb Raider, Deus Ex: Mankind Divided, Baldur’s Gate 3, and Civilization 6. Games were run through benchmark suites where available, recorded games in the case of Dota 2, or consistent camera movements in the case of BG3. Window Captures were compared side by side with General Captures where the game was running at 1080p, windowed, at at least 60fps. The resulting encoded videos were compared frame-by-frame to analyze framerate differences. Activity Monitor was used to track CPU, GPU, and RAM utilization during the tests.
Demo videos consist of two shots composited in OBS. Each scene shows a game captured using either a Window Capture or General Capture as noted, and a cropped General Capture of the Activity Monitor at time of capture.
### Types of Changes
- Bug fix (non-breaking change which fixes an issue)
- Performance enhancement (non-breaking change which improves efficiency)
- Code cleanup (non-breaking change which makes code smaller or more readable)
#### Summary of Changes
- `libobs-opengl/gl-cocoa.m` - Removed a requirement that frames sent to `gl-cocoa` be the same dimensions as the original frame. This is a requirement to support resizing windows. Without this change, window streams would freeze if the window was ever resized. Testing seems to show that this change does not introduce any issues.
- `plugins/mac-capture/CMakeLists.txt:15,16` - Clean up consistency of tabs/spaces in `include_directories`.
- `plugins/mac-capture/CMakeLists.txt:28` - Inclusion of new capture into compile list.
- `plugins/mac-capture/CMakeLists.txt:42,53` - Inclusion of new libraries needed for General Capture. `CoreVideo` and `CoreMedia` are needed for managing the `CVSampleBuffer` format of source frame buffers. `ScreenCaptureKit` enables the new capture type. Include `ScreenCaptureKit` as `weak_framework` to support compiling on pre-12.3 systems.
- `plugins/mac-capture/mac-display-capture.m:621` - The `__MAC_10_15` define may not exist on devices running < 10.15, resulting in this section not being appropriately avoided during compilation. Swapping to `101500` should work for all versions of macOS.
- `plugins/mac-capture/mac-general-capture.m` - Inclusion of new capture type. A significant amount of code is similar to existing code in `mac-display-capture.m`, intentionally.
- `plugins/mac-capture/mac-general-capture.m:4,6` - Availability check for weak linking.
- `plugins/mac-capture/mac-general-capture.m:8,10` - Blocking compilation on 12.3 and later. Since the compilation block is here, availability is left unchecked for each call. The warning is ignored for this file to reduce noise of warnings. Future additions to this plugin type will need attention paid to potential unguarded checks in availability after 12.3.
- `plugins/mac-capture/mac-general-capture.m:164,211` - Given the different availability tools provided by `ScreenCaptureKit`, matches must be generated from the way OBS stores windows/displays to the format `ScreenCaptureKit` provides.
- `plugins/mac-capture/mac-general-capture.m:296,312` - Initial fetch of available capture targets when plugin created. Fetches are also run whenever the plugin properties are opened to ensure accurate listings in the properties.
### Future Work
- Localization of General Capture not included in PR.
- OBS might want to exclude itself from captures (or provide an option to do so) to avoid the “Hall of Mirrors” problem of bringing OBS within capture bounds to edit.
- “General Capture” UI is simplistic to duplicate whats already present, we’d expect the project to revisit how it exposes this mode
- Areas of code reuse are intentional to ensure familiarity with future maintainers and current implementations. If the project desires abstractions to reduce duplication, that should be discussed.
### Checklist:
- [x] My code has been run through clang-format.
- [x] I have read the contributing document.
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

RytoEX commented:
Does this need to be rebased on #5155 or can it be rebased on git master?

Also, how much of this conflicts with existing efforts such as #5680 and #5617 and others in the PR queue?

Developer-Ecosystem-Engineering commented:
> Does this need to be based on #5155 or can it be rebased on git master?
> 

We were asked to base it on 5155, as it does work on Apple silicon and x86. 5155 is the work to enable Apple silicon.

> Also, how much of this conflicts with existing efforts such as #5680 and #5617 and others in the PR queue?

Those PRs are focused on the existing capture type, this was intentionally put under a new capture type so as to not cause conflicts. There could be reasons a project might want to use `CGWindowListCreate` as a capture mode (namely, this will only work on macOS 12.3 or later, and the existing method works much further back)

gxalpha commented:
> plugins/mac-capture/mac-display-capture.m:621 - The __MAC_10_15 define may not exist on devices running < 10.15, resulting in this section not being appropriately avoided during compilation. Swapping to 101500 should work for all versions of macOS.

Since we're only really supporting building on higher than that, I personally think that shouldn't be an issue (and it hasn't been so far)

> OBS might want to exclude itself from captures (or provide an option to do so) to avoid the “Hall of Mirrors” problem of bringing OBS within capture bounds to edit.

Such an option was recently merged for Windows, I would probably be good to use that settings. See #5698

Developer-Ecosystem-Engineering commented:
> > plugins/mac-capture/mac-display-capture.m:621 - The __MAC_10_15 define may not exist on devices running < 10.15, resulting in this section not being appropriately avoided during compilation. Swapping to 101500 should work for all versions of macOS.
> 
> Since we're only really supporting building on higher than that, I personally think that shouldn't be an issue (and it hasn't been so far)
> 
> > OBS might want to exclude itself from captures (or provide an option to do so) to avoid the “Hall of Mirrors” problem of bringing OBS within capture bounds to edit.
> 
> Such an option was recently merged for Windows, I would probably be good to use that settings. See #5698

Something the project should discuss further. The option would only work in this new capture mode, not the other two. The setting today exists in Preferences -> General -> “Hide OBS windows from screen capture”.  Checking it on macOS would mean with this PR if we added support, it would be hidden in one capture mode but not the other two.  That may be better left for a different PR once its decided how to best integrate the modes and settings. The actual implementation of hiding it should be relatively straight forward.


> Does this need to be based on #5155 or can it be rebased on git master?
We can rebase on master if preferred, we had originally done it that way, its not an issue if you'd like to swap back.

itscsk111165 reviewed (commented):
Looks great!

timthelion commented on UI/window-basic-main.cpp:
```diff
@@ -3621,7 +3635,7 @@ bool OBSBasic::QueryRemoveSource(obs_source_t *source)
 #define UPDATE_CHECK_INTERVAL (60 * 60 * 24 * 4) /* 4 days */
 
-#ifdef UPDATE_SPARKLE
+#if defined(ENABLE_SPARKLE_UPDATER)
```
I'm not related to the project at all, but this appears to be a non-semantic refactor, and non-semantic refactors should be avoided in functional/semantic PRs because they tend to make the diff larger and the changes therefore much harder to assess.

RytoEX replied:
As [previously pointed out](https://github.com/obsproject/obs-studio/pull/5875#issuecomment-1023529237), this PR is based on PR #5155, which is the source of changes like this.

timthelion commented on libobs/obs-module.c:
```diff
@@ -357,9 +357,17 @@ static bool parse_binary_from_directory(struct dstr *parsed_bin_path,
 	dstr_cat(parsed_bin_path, file);
+#ifdef __APPLE__
+	if (!os_file_exists(parsed_bin_path->array)) {
+		dstr_cat(parsed_bin_path, ".so");
```
Is the duplication on line 370 intentional?

RytoEX replied:
This file is not modified in this PR. It may have been modified as part of one of its dependent PRs that has since been merged.

timthelion commented on libobs/obs-nix.c:
```diff
@@ -54,16 +54,15 @@ const char *get_module_extension(void)
-	FLATPAK_PLUGIN_PATH "/share/obs/obs-plugins/%module%",
-};
+	OBS_DATA_PATH "/obs-plugins/%module%",
+	FLATPAK_PLUGIN_PATH "/share/obs/obs-plugins/%module%"};
```
This formatting change appears to be a typo. To be consistent with the static declaration above the curly bracket should be on its own line.

timthelion replied:
I think github is broken, these should not be marked as "Outdated"...

RytoEX replied:
This file is not modified in this PR. It may have been modified as part of one of its dependent PRs that has since been merged.

RytoEX commented:
> We were asked to base it on 5155, as it does work on Apple silicon and x86. 5155 is the work to enable Apple silicon.

I'm fine leaving it based on #5155, I just wanted to make sure it was intentional.  In that case, I'll probably convert this to a Draft PR as it requires 5155 to be merged before we'll take action here.

norihiro commented:
The commit message should follow this rule:
> a full description of the commit, wrapped to 72 columns max

And, I'd like to suggest to have more summarized commit message. Maybe just the description part would be ok.

integerrr commented:
I somehow couldn't get the general capture to work on my build of this branch, is there any specific build instructions? 
Note I had to amend the dep hashes as per d7be35851a7e87040d8bea47c1d32eac8348186f manually to get past installing dependencies

thbaja commented:
@Developer-Ecosystem-Engineering 
> Display Capture uses CGDisplayStream which provides efficient access to the screen buffer. Unfortunately, that means that operations like cursor removal are not possible.

Isn't that what the [`showCursor`](https://developer.apple.com/documentation/coregraphics/cgdisplaystream/1454924-showcursor) property is for?

The option of excluding other applications/windows is a very nice addition compared to `CGDisplayStream` though.

Developer-Ecosystem-Engineering commented:
> @Developer-Ecosystem-Engineering
> 
> > Display Capture uses CGDisplayStream which provides efficient access to the screen buffer. Unfortunately, that means that operations like cursor removal are not possible.
> 
> Isn't that what the [`showCursor`](https://developer.apple.com/documentation/coregraphics/cgdisplaystream/1454924-showcursor) property is for?
> 
> The option of excluding other applications/windows is a very nice addition compared to `CGDisplayStream` though.

Yes, you are correct, I've updated the language. Window capture can now _also_ have cursor inclusion/exclusion, like CGDisplayStream previously provided, as you point out!

gxalpha commented:
@Developer-Ecosystem-Engineering just for your info, seeing that you just merged the universal-branch into this PR, the universal PR was merged into master a few days ago, so this could get rebased onto master now.

Developer-Ecosystem-Engineering commented:
> @Developer-Ecosystem-Engineering just for your info, seeing that you just merged the universal-branch into this PR, the universal PR was merged into master a few days ago, so this could get rebased onto master now.

Yeah, the plan is to rebase on master, working on it!

wookayin commented:
Another FYI, direct rebasing would give quite a mess and conflicts because this PR didn't have the correct base branch in the beginning. I was able to make a successful build after cherry-picking the two commits only ("mac-capture:  Add support ..." and "Update to use delegates...") without any conflicts, which could be even squashed, on top of the latest master branch. BTW thank you for adding the amazing feature!

norihiro commented:
Commit message should be wrapped to 72 columns max as the guideline describes.
https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst#commit-guidelines

jcm93 commented:
I have been streaming and recording with General Capture for a few weeks now, very pleased with the results. No outstanding glitches to speak of and solid performance. It is slightly funky with UI integration, given that it is multiple sources (display/window/application capture) in one, but minor UX notwithstanding it works great. I imagine the specific capture types could be separated out later if need be.

I have noticed one minor issue where a General Capture window source will not automatically reacquire its window if OBS is relaunched and the window is in a separate display and Space, but this is quite small.

Developer-Ecosystem-Engineering commented:
> I have noticed one minor issue where a General Capture window source will not automatically reacquire its window if OBS is relaunched and the window is in a separate display and Space, but this is quite small.

Different captures have different behaviors, some follow the app across all displays, other are anchored to the display it starts on. In this scenario I believe you'd want an "App capture", which should follow the app window regardless of the display or space it appears on.

jcm93 commented:
> Different captures have different behaviors, some follow the app across all displays, other are anchored to the display it starts on. In this scenario I believe you'd want an "App capture", which should follow the app window regardless of the display or space it appears on.

I was slightly unclear. Specific steps:

1. "Displays have separate Spaces" is checked in sys prefs.
2. Window being captured is located on Display 1, Space 2. User focus is Display 1, Space 1. OBS is located on Display 2, Space 1.
3. Relaunch OBS. The captured window will not be reacquired, despite not moving.

Squinting at this harder, I'm guessing this may be intended behavior, since General Capture seems to be generally unaware of windows in separate Spaces on other displays unless they are focused, so this may simply be how the ScreenCaptureKit API and Spaces work.

Also, FWIW, while writing these steps, I ran into a strange issue with a Safari window capture doubling/halving in size/scale in OBS when the mouse cursor moved between displays (?), but I was unable to reliably reproduce it.

CarrotCodes commented:
I got a chance to try this out today on an M1 Ultra machine.

Built from the branch using `./ci/build-macos.sh`. I'd echo the above comment about the source UI being a bit confusing, but it's already noted in the PR description and it is functional. In particular, some of the fields don't seem relevant in all modes (like "application") but I'm sure it's easy enough to fix in a UX improvement pass.

I streamed for about 4 hours, didn't notice any issues (other than Twitch having some slightly odd keyframe issues, but I'm pretty sure it's unrelated), and performance was much improved - great work 👍

jcm93 commented:
I reproduced the doubling/halving in size thing.

If a window General Capture is clipped at the edge of my display, moving the cursor between areas makes the window General Capture freak out.

This seems to occur regardless of whether "Displays have separate Spaces" is on or off. I am sure it has something to do with my non-Retina external display, but even if the window does not actually clip into a non-Retina area, the issue occurs. Attached is a video demonstrating. OBS is the app focused in the video, but the issue also occurs if Safari is the focused application.

https://user-images.githubusercontent.com/6864788/163622604-d2fd451d-b256-456e-9f62-aebe5e8b027f.mp4

jcm93 commented:
Also, on the topic of Retina/non-Retina:

On my system, General Capture using the display or application modes seem to be capturing at downscaled rather than full display resolution. Only the window mode seems to capture at full resolution. Is there something I'm missing here? This is without an external display plugged in or anything of that sort, hardware is an M1 Max MBP.

Attached: vanilla display capture, display General Capture, application General Capture.

<img width="372" alt="Screen Shot 2022-04-15 at 3 26 16 PM" src="https://user-images.githubusercontent.com/6864788/163628951-f2e23447-152e-45cb-9bb3-e1d37bae80b1.png">

carlosqsilva commented:
On my system Application Capture only works with some system apps (Finder, System Preference, Safari), all third party apps i tested did not worked 

https://user-images.githubusercontent.com/19891059/163696309-e8a0878b-bbd9-46cb-9422-eaa02cfbf14b.mp4

Display and Window capture works great 👍

jcm93 commented:
> On my system Application Capture only works with some system apps (Finder, System Preference, Safari), all third party apps i tested did not worked

I can personally vouch for at least two third-party applications working with Application capture; I have successfully captured Parallels sessions with it as well as Steam. But yes, it appears many do not work correctly or are quite finicky.

Developer-Ecosystem-Engineering commented:
We do not believe this is an App or OBS related issue at the moment. For @carlosqsilva 

>On my system Application Capture only works with some system apps (Finder, System Preference, Safari), all third party apps i tested did not worked

Did you only try selecting applications that were not visible on the selected monitor at the time the capture started? Its unclear if what you and @jcm93 are seeing are "OBS" issues or issues with ScreenCaptureKit. 

We wouldn't expect some applications to be finicky or not work correctly. If you do discover an issue like that, those are likely outside of OBS, so a [Feedback](https://developer.apple.com/bug-reporting/) to Apple to investigate it would be appreciated!

jcm93 commented:
Here's a quick poll of everything I have running now. With no external displays connected, and all applications listed visible on the main display at the time of being selected, Application general capture doesn't capture any of the following (while window general capture works fine):

Dota 2
TextMate
Blender
Discord
Logic Pro
Messages
Music

For these, no windows are visible, though the cursor is still captured.

Application capture does successfully capture:

Terminal
Xcode
Safari
Finder
Steam
Notes

I am guessing this points to being an issue with ScreenCaptureKit? If if there's else anything I can do concretely verify that it isn't OBS/General Capture, feel free to let me know.

Developer-Ecosystem-Engineering commented:
Heya @jcm93, @carlosqsilva  thanks for your persistence! The oddity of the matrix (Steam works, Dota 2 doesn't, Notes does, Music doesn't) had us looking with a fine toothed comb! We believe we've found at least one issue, changes pending. This would go into OBS.

RytoEX commented:
With #6387 now merged, CI is now using the GitHub Actions macos-12 image using Xcode 13.3 and the macOS 12.3 SDK with ScreenCaptureKit. Once this is rebased, CI should produce builds for this PR that function as described on macOS 12.3+. I will remove the draft status from this PR once it has been rebased.

Developer-Ecosystem-Engineering commented:
Thank you all for your findings. This latest change includes the following

- General Capture & Application Capture not showing all applications should be resolved
- Retina/non retina quality/downscaling should be resolved
- Minor modifications to improve resizing behaviors
- Rebase as requested by @RytoEX 

Absent additional feedback this should be good to go.

RytoEX commented:
Removed the draft status. Build artifacts are currently here:
* [M1 (ARM64)](https://github.com/obsproject/obs-studio/suites/6379459701/artifacts/231521860)
* [Intel (x86_64)](https://github.com/obsproject/obs-studio/suites/6379459701/artifacts/231521861)

Further testing and feedback are welcome.

integerrr commented:
Some full screen games are still slightly finicky to setup application capture properly.
Currently I still have to:
1. switch to the fullscreen game
2. (on a separate monitor) change the capture source to display capture
3. configure the source as display capture to capture the game (application dropdown doesnt show the game when you're not in the game i.e. on another "desktop")
4. save the source config
5. change the config to application capture again

I think this breaks every time you restart the game or OBS, not the end of the world, just a bit annoying. Thank you for your work tho!

robotspacer commented:
I’ve used a previous build for a couple of Twitch streams without any significant issues. Performance is significantly improved over the existing capture methods, and it also fixes a flickering issue I’ve had with Display Capture. The latest builds fixes Application Capture for the apps I noticed it with.

I’ve noticed one minor regression compared to the existing “Window Capture” source. The issue is still present in the latest build.

Steps to reproduce:
- Add a “General Capture” source
- Set the method to “Window Capture”
- Select a window to capture
- Quit both OBS and the app/window being captured (as one might do between uses)
- Open OBS (before opening the window that was being captured)
- Open the app/window being captured

Expected behavior:
OBS should start capturing the window after it appears. This is the behavior when using the existing “Window Capture” source.

Actual behavior:
The source remains blank. It can be fixed by restarting OBS, or by editing the source’s properties and reselecting the window. Editing the properties is a bit cumbersome, because the window now appears twice in the pop-up menu, but the other item is disabled. I have to select some other random window first, at which point the duplicate item disappears. Then I’m able to switch back to the window I’m trying to capture.

A similar issue happens if I close the window I’m capturing while OBS is running. OBS will continue to show the window as it was when it was closed, which is fine. But when the window is re-opened, the capture doesn’t resume. At this point it’s basically the same problem as above—I need to restart OBS or reselect the window to fix it.

Developer-Ecosystem-Engineering commented:
> Some full screen games are still slightly finicky to setup application capture properly. Currently I still have to:
> 
> 1. switch to the fullscreen game
> 2. (on a separate monitor) change the capture source to display capture
> 3. configure the source as display capture to capture the game (application dropdown doesnt show the game when you're not in the game i.e. on another "desktop")
> 4. save the source config
> 5. change the config to application capture again
> 
> I think this breaks every time you restart the game or OBS, not the end of the world, just a bit annoying. Thank you for your work tho!

It would be helpful to know what specific games you are referencing. Games can take several different approaches to implementing what a user experiences as "full screen".  At the content capture level its useful for us to understand. 

Separately, what's driving the swap in Step 3 to Step 5 (going from display to app capture)?

integerrr commented:
> > Some full screen games are still slightly finicky to setup application capture properly. Currently I still have to:
> > 
> > 1. switch to the fullscreen game
> > 2. (on a separate monitor) change the capture source to display capture
> > 3. configure the source as display capture to capture the game (application dropdown doesnt show the game when you're not in the game i.e. on another "desktop")
> > 4. save the source config
> > 5. change the config to application capture again
> > 
> > I think this breaks every time you restart the game or OBS, not the end of the world, just a bit annoying. Thank you for your work tho!
> 
> It would be helpful to know what specific games you are referencing. Games can take several different approaches to implementing what a user experiences as "full screen". At the content capture level its useful for us to understand.
> 
> Separately, what's driving the swap in Step 3 to Step 5 (going from display to app capture)?
-------

Sorry for the bad explanation in the last post. 

The game concerned here is called [osu!](https://github.com/ppy/osu) which was rewritten and uses SDL to handle screen modes/game window I believe. In my example, the game was run in "borderless" mode which is similar to how e.g. native apps run in fullscreen, as opposed to "(exclusive) fullscreen" mode which does not allow app switching as well as displaying anything in other monitors in macos i.e. the 2nd monitor will be blacked out.

Simply, if I dont start obs with the game already open and focused (which lives on a second "desktop", see below)
> <img width="394" alt="image" src="https://user-images.githubusercontent.com/7279624/167036096-1782eb8d-7189-4ff5-a143-04ce8e44f6a2.png">

The capture source will not be able to find the game and will set itself to capture another app instead, I would then need to fix the source following the steps from the timestamp 1:02 in the video below. I hope the video will help convey what I am trying to say here.

https://user-images.githubusercontent.com/7279624/167036446-0a4a12e5-ac61-4f96-97ae-7957b83e34c0.mp4

Developer-Ecosystem-Engineering commented:
> The capture source will not be able to find the game and will set itself to capture another app instead, I would then need to fix the source following the steps from the timestamp 1:02 in the video below. I hope the video will help convey what I am trying to say here.
> 

Thanks it was very helpful!  What we see is expected behavior.  In that specific scenario for the app to appear
https://github.com/obsproject/obs-studio/blob/651ae554bffe9b8aa1c6047f3afce42c5d4fbeac/plugins/mac-capture/mac-general-capture.m#L466 would need to be set to false.   That will result in many more entries appearing in the list of applications which may be confusing to the average user, but would result in your app appearing when its offscreen.

It would be a good future improvement to filter the exhaustive list to only include entires that make sense for the projects users.

ldenoue commented:
@Developer-Ecosystem-Engineering I noticed a strange behavior when capturing a single window with SCContentFilter(desktopIndependentWindow: window): parts of that window that fall outside of the display are black. Is it by design or a bug? Apps such as Zoom and Teams now have this behavior when sharing a single window (which many people do). OSB users might as well.

Developer-Ecosystem-Engineering commented:
> @Developer-Ecosystem-Engineering I noticed a strange behavior when capturing a single window with SCContentFilter(desktopIndependentWindow: window): parts of that window that fall outside of the display are black. Is it by design or a bug? Apps such as Zoom and Teams now have this behavior when sharing a single window (which many people do). OSB users might as well. The CGWindowListCreateImage API returns the complete window image, which is the expected behavior (for me at least :) ) See the video demonstration that uses ScreenCaptureKit
> 
>  screencapturekit-crops-window-sharing.mp4 
> 
> .

Hi @ldenoue,

On current macOS 12.x you either have a canvas the size of the display, or a canvas the size of the window as visible on the display. This doesn't sound related to the OBS Project and would be a better fit for a [Feedback](https://developer.apple.com/bug-reporting/) about ScreenCaptureKit, thanks!

ldenoue commented:
@Developer-Ecosystem-Engineering unfortunately 12.4 hasn't fixed the issue for me (tested on Zoom, Teams and my own CaptureSample ScreenCaptureKit).
I "solved" the issue now with a virtual camera that lets my users select it to show the Screegle Window https://www.appblit.com/screegle#plugin

PatTheMav commented:
I did a UX review on this, I came away with a few things:

* The UI should not expose a Window and Application selection at the same time - the first will only have an effect if "Window Capture" method is selected, the latter only for "Application" capture. Having both available at the same time confuses users, because one of the dropdown fields will become ineffective (and both become ineffective when in "Display capture" mode).
* Application capture seems to capture just the textures of a given application as placed on the main screen and space. It is not possible to select an application's window on another space (if an app's windows are distributed that way). Also users will have to manually crop the visible area to just the application window, which will not follow size changes of that application window and could also lead to blurry captures depending on the scaling filter.
* The "Show windows with empty names" checkbox is visible and selectable, but doesn't seem to have any effect. Please remove the checkbox, or implement the associated functionality.
* Changes to the list of available windows seems to be updated only after selecting a window. New windows like Safari's PIP video player are shown in the list, but no output is visible after selecting it..
* Error messages are printed using OBS' `blog` method, but without the common prefix in square brackets (e.g. `[mac-capture]`). Please check other plugins for how the automatically prefix all log messages with their plugin name.
* I'd rather rename this source "macOS ScreenCapture" or something that more clearly communicates that we use built-in functionality - "general capture" is not obvious enough IMO.
* In my tests OpenEMU couldn't be captured at all - its windows neither appeared in the window list nor did it appear in the applications list.

Technically the capture seems to work great performance-wise, but from a UX POV the property window needs to be more reactive to the type of capture selected and Application Capture needs a rethink or additional features to be less confusing to users.

For the non-macOS users, here's a screenshot of the properties window:

<img width="940" alt="Bildschirmfoto_2022-06-04_um_15 23 37" src="https://user-images.githubusercontent.com/9436503/172013460-81a77e05-fad5-4f44-b9cb-e0f44e2a7d3b.png">

@gxalpha, @Fenrirthviti, @RytoEX, any thoughts from your end?

---

The prefixed logging can be implemented by aliasing `blog`, e.g. like this:

```
#define PLUGIN_NAME "mac-screencapture"

#define blog(level, msg, ...) \
    blog(level, "[" PLUGIN_NAME "] " msg, ##__VA_ARGS__)
```

Developer-Ecosystem-Engineering commented:
Hi @PatTheMav, thanks so much for the review! Replies inline.

> I did a UX review on this, I came away with a few things:
> 
> * The UI should not expose a Window and Application selection at the same time - the first will only have an effect if "Window Capture" method is selected, the latter only for "Application" capture. Having both available at the same time confuses users, because one of the dropdown fields will become ineffective (and both become ineffective when in "Display capture" mode).

As the community uses the feature more, how to best express these various capture types will likely go through some iteration. That will need to happen in a new PR. 

> * Application capture seems to capture just the textures of a given application as placed
 on the main screen and space. It is not possible to select an application's window on another space (if an app's windows are distributed that way). Also users will have to manually crop the visible area to just the application window, which will not follow size changes of that application window and could also lead to blurry captures depending on the scaling filter.

This is the expected behavior for ScreenCaptureKit. 

The anticipated use case for this type of capture is something like a tutorial video where showing application positioning, additional windows, and pop-ups would be necessary for the flow. This means that "Application Capture" is dependent on the selected "Display."  If you are looking to crop your capture to a single application window, the existing Window Capture fills that need today.

> * The "Show windows with empty names" checkbox is visible and selectable, but doesn't seem to have any effect. Please remove the checkbox, or implement the associated functionality.

Removed

> * Changes to the list of available windows seems to be updated only after selecting a window. 

Changes to the list of available windows only happens when the settings for the instance of the ScreenCaptureKit-based capture is opened. This is identical to the current behavior for "Window Capture" and was done to emulate that existing behavior

> New windows like Safari's PIP video player are shown in the list, but no output is visible after selecting it..

We don't see this issue, could you provide steps to reproduce? Might be suited to a different PR/New Issue.
 
- Using MacOS 12.4
- Create an instance of the ScreenCaptureKit-based capture
- Close capture preferences
- Start a PIP video (tested with a YouTube video)
- Open SCKit-based capture preferences
- Method: Window Capture
- Window: [PIPAgent] pip

> * Error messages are printed using OBS' `blog` method, but without the common prefix in square brackets (e.g. `[mac-capture]`). Please check other plugins for how the automatically prefix all log messages with their plugin name.

Added `MACCAP_LOG` and `MACCAP_ERROR` to resolve this issue. 

> * I'd rather rename this source "macOS ScreenCapture" or something that more clearly communicates that we use built-in functionality - "general capture" is not obvious enough IMO.

Changed. File is now `mac-screen-capture.m`, internal references are now `ScreenCapture*` and `screen_capture_* `as necessary. All instances of `dc` swapped to `sc.`

> * In my tests OpenEMU couldn't be captured at all - its windows neither appeared in the window list nor did it appear in the applications list.

We don't see this issue, could you provide steps to reproduce? Might be suited to a different PR/New Issue. 

- Using macOS 12.4
- Launch OpenEMU
- Create an instance of the ScreenCaptureKit-based capture
- OpenEMU is present in both the application and window list while viewing the library.

> * Error 3808 in `stopCaptureWithCompletionHandler` ("stream already ended") should also be caught gracefully instead of printing these warnings/errors in the OBS log if this is an unfortunate but possible edge case.

Updated to explicitly prevent the plugin from logging for this error code

PatTheMav commented:
In that case we'd have to force-push the UI changes ourselves because as-is the UX is (at least in my opinion) not intuitive enough for our users to be merged. I'll see what I can do.

While testing the changes, I encountered a deadlock when the source is added for the first time and screen recording permissions need to be granted:

* Entered the preferences window
* OBS was already present (prior development build with different `cdhash` not matching the currently executed build)
* Removed OBS from the list
* Closed the preferences window

OBS is now stuck in the "create new source" step - according to Xcode the app hangs in this call: https://github.com/Developer-Ecosystem-Engineering/obs-studio/blob/f63f76073fcc094ed2d2fc5aac6c8c37341e3932/plugins/mac-capture/mac-screen-capture.m#L84

PatTheMav commented:
@Developer-Ecosystem-Engineering thank you for the updates, reviewing them now.

> > New windows like Safari's PIP video player are shown in the list, but no output is visible after selecting it..
> 
> We don't see this issue, could you provide steps to reproduce? Might be suited to a different PR/New Issue.

I checked this issue again, the core issue is that the window list is not updated while the properties window is open - if I collapse the PIP window and extend it again, the list entry will not be updated.

I dunno if there's a better fix for this, but that's more an issue with the list of windows staying static.

> > * In my tests OpenEMU couldn't be captured at all - its windows neither appeared in the window list nor did it appear in the applications list.

I tested this again as well, OpenEMU can be captured, as long as the game window is not made into a fullscreen window. As soon as it's full screen, it disappears.

For that reason I'd rather have a functioning "show hidden windows" checkbox, if that would make the full screen window appear as this seems to be common use case by other apps, games, and game frameworks and would defeat the purpose of a new and improved capture plugin if it cannot capture the most common use case for it (full screen games).

PatTheMav commented:
Further notes after review:

* Please add `update_window(&sc->window, settings);` to `screen_capture_create` to populate the window field with the `window` value from settings. Otherwise the plugin will not correctly start up in "window capture" mode
* Please add `obs_data_set_default_int(settings, "window", kCGNullWindowID);` to initialise settings a proper default window id

I've implemented the desired functionality for the dropdown to show/hide the appropriate selection of display/window/applications. I'll reach out off-thread as to how to get those incorporated.

PatTheMav commented:
@Developer-Ecosystem-Engineering - thank you for the quick turnarounds on this - I ran another set of tests with the newest commits:

* Got a screen capture permission prompt before the main OBS window came up (that is with the source present in the scene on program launch)
* Removed the OBS entry from permissions
* OBS stays stuck with bouncing app symbol in dock, application is deadlocked

This needs to be handled more gracefully - missing permissions cannot deadlock/block the app.

* I relaunched the app, added the macOS capture source
* Permission prompt came up, I added the permission
* Did _not_ restart the app
* Was able to add the source, but permissions button is disabled

In this situation the source was added, but because it didn't have the necessary permissions, the source code not be created:

```
21:19:23.313: [ mac-screencapture ]: screen_capture_properties: Failed to get shareable content with error (null)
21:19:23.313: 
21:19:23.314: [ mac-screencapture ]: init_screen_stream: Failed to start capture with error (null)
21:19:23.314: 
21:19:23.316: [ mac-screencapture ]: destroy_screen_stream: Failed to stop stream with error (null)
21:19:23.316: 
21:19:23.316: Failed to create source 'macOS ScreenCapture'!
21:19:23.320: User added source 'macOS ScreenCapture' (screen_capture) to scene 'Scene'
```

This also leaves the app in a weird state. I dunno if this needs to be more graceful, given that a user should prefer relaunching the app when giving the permission.

Yet another restart got OBS and the source back in working order

Next, when switching between window and application capture, I encountered the bug that when a full screen window was selected for window capture (using the new "show full screen windows" setting), the window capture was only taking up a fraction of the source output.

Switching between windows restored the "full screen" aspect of it (read: the captured window was captured at its actual size).

Developer-Ecosystem-Engineering commented:
> @Developer-Ecosystem-Engineering - thank you for the quick turnarounds on this - I ran another set of tests with the newest commits:
> 

Thanks, reviewing & will reply with latest!

Developer-Ecosystem-Engineering commented:
> Further notes after review:
> 
> * Please add `update_window(&sc->window, settings);` to `screen_capture_create` to populate the window field with the `window` value from settings. Otherwise the plugin will not correctly start up in "window capture" mode
> * Please add `obs_data_set_default_int(settings, "window", kCGNullWindowID);` to initialise settings a proper default window id
> 

These are done in the latest push

> OBS stays stuck with bouncing app symbol in dock, application is deadlocked

This is resolved in the latest push

> I encountered the bug that when a full screen window was selected for window capture (using the new "show full screen windows" setting), I then switch to application capture and back to window capture, the window capture output was only taking up a fraction of the source output.

This is resolved in the latest push

> For that reason I'd rather have a functioning "show hidden windows" checkbox

This is resolved in the latest push

Should be all set thanks for the review!

PatTheMav commented:
Can confirm the reported issues are gone.

A small nitpick I have is that upon enabling the "Show windows with empty names" function the entire properties windows "resets", that is if I selected window capture prior to ticking the box, it resets to display capture. When selecting window capture, the old window is not necessarily selected again as well.

I'd be fine with merging this as-is once the bug reported below is fixed though, as we'll add a UX/UI pass commit on top of this anyway.

@Developer-Ecosystem-Engineering I did find a new bug - when "Show cursor" is disabled, when capturing "PIPAgent", I do get the cursor in the capture the moment I stop moving the mouse. Once I move it, the cursor disappears, but reappears once movement is finished.

Developer-Ecosystem-Engineering commented:
> Can confirm the reported issues are gone.
> 
> A small nitpick I have is that upon enabling the "Show windows with empty names" function the entire properties windows "resets", that is if I selected window capture prior to ticking the box, it resets to display capture. When selecting window capture, the old window is not necessarily selected again as well.
> 

Will take a look thanks!

> I'd be fine with merging this as-is once the bug reported below is fixed though, as we'll add a UX/UI pass commit on top of this anyway.
> 


> @Developer-Ecosystem-Engineering I did find a new bug - when "Show cursor" is disabled, when capturing "PIPAgent" via application capture(!), I do get the cursor in the capture the moment I stop moving the mouse. Once I move it, the cursor disappears, but reappears once movement is finished.

What OS build?

PatTheMav commented:
> > @Developer-Ecosystem-Engineering I did find a new bug - when "Show cursor" is disabled, when capturing "PIPAgent" via application capture(!), I do get the cursor in the capture the moment I stop moving the mouse. Once I move it, the cursor disappears, but reappears once movement is finished.
> 
> What OS build?

ProductName:	macOS
ProductVersion:	12.4
BuildVersion:	21F79

HiroiSekai commented:
I'm absolutely terrible at GitHub, how can I acquire the latest push to test? Thanks :)

HiroiSekai commented:
Thanks for working on this! I thought it might be helpful if someone tested it going forward on the currently running macOS Ventura beta builds. I know it's introductory and I'm only listing stuff here as notes. So far:

- I noticed the resize docks interface had been updated, I still have the issue of them being really fidgety and not responding to resizing easily.

- The dock options that come from logging into your Twitch account in OBS (Twitch Stats, Twitch Chat, etc.) are missing from the original app feature list. I'm just guessing this is because there isn't full Twitch API integration in this build yet?

- Window Capture for me works really great so far! I tried it with Photoshop CC 2022, Google Chrome, and a few of Apple's apps like Music and Notes, all worked properly.


Great work so far, thanks again! I'm still terrible at GitHub though, I found the artifact build on this page but I don't know how to get pushed update builds to test on Mac. Would love some help there.

RytoEX commented:
> Thanks for working on this! I thought it might be helpful if someone tested it going forward on the currently running macOS Ventura beta builds. I know it's introductory and I'm only listing stuff here as notes. So far:
> 
>     * I noticed the resize docks interface had been updated, I still have the issue of them being really fidgety and not responding to resizing easily.

This is likely unrelated to anything in this PR.

>     * The dock options that come from logging into your Twitch account in OBS (Twitch Stats, Twitch Chat, etc.) are missing from the original app feature list. I'm just guessing this is because there isn't full Twitch API integration in this build yet?

PR builds do not have service integrations (Twitch, YouTube, etc.).

>     * Window Capture for me works really great so far! I tried it with Photoshop CC 2022, Google Chrome, and a few of Apple's apps like Music and Notes, all worked properly.
> 
> 
> Great work so far, thanks again! I'm still terrible at GitHub though, I found the artifact build on this page but I don't know how to get pushed update builds to test on Mac. Would love some help there.

There is no way to get automatic updates to PR builds.

gxalpha commented:
@HiroiSekai looking at the name, that's not the newest test build of this Pull Request.
I just added a link to the newest test build to the top of the PR description which we'll try to keep updated, you can check back from time to see if there is a newer version.

MisutaaAsriel commented:
Been testing this recently and no issues insofar. It actually seems pretty snappy on my 2017 i5 MacBook Pro. I have no complaints so far. :)

Southpaw1496 commented:
I've been testing the latest build this afternoon, and the performance improvement over Window Capture is fantastic. I did, however, find a a couple of bugs, both of which can be seen in the enclosed video.

The first was a bug which occurred in the old Window Capture system, and is still present with the new implementation. If you're capturing a full-screen window and leave it using the three-finger swipe gesture, its position in the capture source slides around a little, before eventually returning to its proper position.

The second bug is that the mouse cursor appears on the capture even when the app is
underneath another app in windowed mode, or unfocused in full-screen mode. This behaviour does not occur when using Window Capture.

https://user-images.githubusercontent.com/44805409/175789277-f28d4bb2-3ce1-4cd6-a9bd-be3b814ff95f.mp4
